spec/GraphQLOverHTTP.md

```diff
@@ -511,19 +511,16 @@ execution regardless of validation errors.
 
 ## Status Codes
 
-In case of errors that completely prevent the generation of a well-formed
-_GraphQL response_, the server SHOULD respond with the appropriate status code
-depending on the concrete error condition, and MUST NOT respond with a `2xx`
-status code when using the `application/graphql-response+json` media type.
```

Comment:
IMO, there's a tension between the [Body first paragraph](https://graphql.github.io/graphql-over-http/draft/#sec-Body):

```
The body of the server’s response MUST follow the requirements for a GraphQL response
encoded directly in the chosen media type.
```

and that paragraph, which is why I removed it. 

My current understanding is that this paragraph was meant for intermediary servers that may not understand GraphQL. So there are 2 kind of servers:

* GraphQL server
* Intermediary server

```
Client <--> [Intermediary Server] <--> GraphQL Server
```

I'd say this spec should be about GraphQL servers and not Intermediary servers? Or if we want to specify the behaviour of intermediary server, we could do it in a separate section?

Or were there use cases when a GraphQL server is compliant and does not return a well formed GraphQL response?

Reply:
At first glance, it appears the paragraph was intended to address invalid requests.  For example, if the request was not able to be parsed as JSON or had an unrecognized content type.  This paragraph would enforce that a non-200 status code was used.  Whereas if we read just above the 6.1 paragraph you referenced, paragraph 6 starts with:

> When a server receives a well-formed [GraphQL-over-HTTP request](https://graphql.github.io/graphql-over-http/draft/#graphql-over-http-request) ...

Reply:
My gripe is specifically about that part: `In case of errors that completely prevent the generation of a well-formed
_GraphQL response_`. That contradicts 6.1. I understand to be 6.1 more general about the body. Then the status codes are refined below.

 If the request was not able to be parsed as JSON or had an unrecognized content type, I think there are other places in the spec that address this already: 

*  request was not able to be parsed as JSON
  * [application/json case](https://graphql.github.io/graphql-over-http/draft/#sel-FANNLRCAACEBi8K) 
  * [graphql-response+json case](https://graphql.github.io/graphql-over-http/draft/#sel-FANNNRCAACENz5F)
* [unrecognized content type](https://graphql.github.io/graphql-over-http/draft/#sel-EANHJCAACElBzyf)

Reply:
I'm not sure this paragraph contradicts anything, but I'm okay removing it as redundant / described elsewhere.

Reply:
The contradiction IMO is in the spec saying: `the body must always be a well formed GraphQL response`. And then going on to say "In case of errors that prevent the generation of a well-formed _GraphQL response_", which I understand as `if it isn't a well formed GraphQL response, then ...`

How can it **not** be a well formed GraphQL response if it must always be a well formed GraphQL response? 

The only reason I see for the body to not be a well formed GraphQL response is if content negociation fails. But as soon as `application/json` or `graphql-response+json` is selected, my reading is that the body must always be a well formed GraphQL response? 

If content-negociation errors are what's at stake here (i.e. "the errors that prevent the generation of a well formed GraphQL response"), then this case is already handled by 6.1 so I'd remove the sentence. If there are other error cases, I'd love to explicit them cause it's not obvious what they could be.

Reply:
> The body of the server’s response MUST follow the requirements for a GraphQL response encoded directly in the chosen media type.

I think this was poorly worded; perhaps:

> When encoding a GraphQL response, the body of the server’s response MUST follow the requirements in the chosen media type.

Essentially it's saying _when you encode a GraphQL response_ that it _must adhere to the spec_. The other paragraph (the one you've deleted) is saying _if_ you cannot form a valid GraphQL response _then_ you must not use a GraphQL response type.

---

> and MUST NOT respond with a `2xx` status code when using the `application/graphql-response+json` media type.

This is also incorrect; more correct would be:

> and MUST NOT respond with the `application/graphql-response+json` media type.

The `application/graphql-response+json` is only for use with well-formed GraphQL responses, so if you can't form a well-formed GraphQL response you can't use it. I think I was probably thinking something like: `try { return doTheGraphQLStuff() } catch (e) { return { errors: [{message: e.message}] }}` where you could then use the graphql-response+json for these errors; but this _is_ a well-formed GraphQL response so it contradicts the beginning of the sentence.

---

This doesn't really refer to "intermediary" servers as much as it does the hosting environment. For example if you use `nginx` and `lua` to do your GraphQL, but the `lua` process crashes, then this requires that your `nginx` isn't configured to still use the `application/graphql-response+json` response type.

Reply:
> Essentially it's saying when you encode a GraphQL response that it must adhere to the spec. The other paragraph (the one you've deleted) is saying if you cannot form a valid GraphQL response then you must not use a GraphQL response type.

This is what my brain has a really hard time processing  😅 . Thinking in  terms of logic:

proposition P: `content-type == application/graphql-response+json` 
proposition Q: `isWellFormGraphQLResponse(body) == true`

> when you encode a GraphQL response that it must adhere to the spec

I think this is the same as `if P then Q`?

> if you cannot form a valid GraphQL response then you must not use a GraphQL response type.

I understand map that as `if !Q then !P`?

So the second sentence is the contrapositive of the first one and therefore strictly equivalent from a logical point of view? 

If it brings the attention to a specific point I'm fine with it as a non-normative note but from in the spec text itself, I'd expect no redundancy. Or is there some nuance there?

Reply:
> > when you encode a GraphQL response that it must adhere to the spec

> I think this is the same as `if P then Q`?

I don't think so. `P` as you have stated it is that you use a specific media type, but what's being stated is that you are encoding a GraphQL response; it could be in any supported media type.

Reply:
Alright 👍  But I think my initial point that it's not clear whether the "non-valid GraphQL response" is part of the spec or not still holds. 

I put the paragraph back in to keep that PR small and about non-normative client notes. Will open a follow up one for "non-valid GraphQL responses"